### PR TITLE
RHEL Installation Script

### PR DESCRIPTION
**Pull request by slifty**
This script will install a fresh copy of the system on an otherwise un-configured RHEL box; it has been tested on RHEL 7.3.

### CODE REVIEW

File: rhel-install.sh [hunk 1]
```diff
@@ -0,0 +1,77 @@
+#!/bin/bash
```
Comment (jasonaowen) [inline rhel-install.sh]: Probably want to `set -e` to stop on errors, and maybe also `-v` to print shell input lines as they are read.

File: rhel-install.sh [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+sudo yum -y update
+< /dev/urandom tr -dc _A-Z-a-z-0-9 | head -c30 | awk '{print $1"!"}' > pass.txt
+curl -O https://download.postgresql.org/pub/repos/yum/9.2/redhat/rhel-7-x86_64/pgdg-redhat92-9.2-3.noarch.rpm
```
Comment (jasonaowen) [inline rhel-install.sh]: Do we need to be on an old version? [9.2 is only supported until September](https://www.postgresql.org/support/versioning/). If not, I suggest we should use latest (currently 9.6).

File: rhel-install.sh [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+sudo rpm -ivh pgdg-redhat92-9.2-3.noarch.rpm
```
Comment (jasonaowen) [inline rhel-install.sh]: Can we validate the RPM before installing it, such as by using gpg or sha512sum or similar?

+1 to installing from the PostgreSQL YUM repo; that should make security updates easier!

File: rhel-install.sh [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+rm pgdg-redhat92-9.2-3.noarch.rpm
+sudo yum -y install postgresql92-server postgresql92-contrib postgresql92-devel
+sudo /usr/pgsql-9.2/bin/postgresql92-setup initdb
+sudo sed -i 's/ident/md5/g' /var/lib/pgsql/9.2/data/pg_hba.conf
+sudo systemctl enable postgresql-9.2.service
+sudo systemctl start postgresql-9.2.service
+sudo -upostgres psql -c "create user \"psm\" with password '$(< pass.txt)'"
+sudo -upostgres psql -c "create database \"psm\" with owner \"psm\" encoding='utf8' template template0"
```
Comment (jasonaowen) [inline rhel-install.sh]: I think these can be replaced with [createuser](https://www.postgresql.org/docs/current/static/app-createuser.html) and [createdb](https://www.postgresql.org/docs/current/static/app-createdb.html).

File: rhel-install.sh [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+sudo yum -y install git
+git clone https://github.com/OpenTechStrategies/psm.git
+sudo yum -y install java-1.8.0-openjdk
+sudo yum -y install ant
+sudo yum-config-manager --enable rhui-REGION-rhel-server-optional
+sudo yum -y install gcc mysql-devel ruby-devel rubygems
+sudo yum -y groupinstall 'Development Tools'
+sudo yum -y install sqlite-devel
```
Comment (jasonaowen) [inline rhel-install.sh]: It might be faster to merge these into one `yum install` command, so it doesn't have to check repolists every time.

File: rhel-install.sh [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  --xa-datasource-class=org.postgresql.xa.PGXADataSource \
+  --valid-connection-checker-class-name=org.jboss.jca.adapters.jdbc.extensions.postgres.PostgreSQLValidConnectionChecker \
+  --exception-sorter-class-name=org.jboss.jca.adapters.jdbc.extensions.postgres.PostgreSQLExceptionSorter \
+  --enabled=true \
+  --use-ccm=true \
+  --background-validation=true \
+  --user-name=psm \
+  --password=$(< pass.txt) \
+  --xa-datasource-properties=ServerName=localhost,PortNumber=5432,DatabaseName=psm
+EOF
```
Comment (jasonaowen) [inline rhel-install.sh]: The `jboss-cli` commands to create mail session, message queue, and data sources could all be merged together, unless you think that makes it less readable.

Comment (jvasile) [inline rhel-install.sh]: I like them apart for readability.

File: rhel-install.sh (hunk 3 above)
Comment (jvasile) [inline rhel-install.sh]: I just pushed a change to this branch that checks the sha1 of the three curl'd items: [commit ff100586a547997b51881602c207e982911940f4](https://github.com/OpenTechStrategies/psm/pull/42/commits/ff100586a547997b51881602c207e982911940f4)

File: rhel-install.sh [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+rm wildfly-10.1.0.Final.tar.gz
+
+cat pass.txt | awk '{print "psm "$1}'>> script.txt
+eval ./wildfly-10.1.0.Final/bin/add-user.sh "$(< script.txt)"
+rm script.txt
+./wildfly-10.1.0.Final/bin/standalone.sh \
+    -c standalone-full.xml \
+    -b 0.0.0.0 \
+    -bmanagement 0.0.0.0 &
+wait_for_jboss
```
Comment (jvasile) [inline rhel-install.sh]: Generally, if I want a server to be up, I try to be actively checking for that rather than sleeping and assuming.  You might have a good reason for sleeping, though, so I'm highlighting the change and I put it in a separate patch: [commit 78fac27f397bb34241b741a559e80c21b4afc26d](https://github.com/OpenTechStrategies/psm/pull/42/commits/78fac27f397bb34241b741a559e80c21b4afc26d)

File: rhel-install.sh (hunk 1 above)
Comment (slifty) [inline rhel-install.sh]: set -e caused the script as written to exit in a few place on non-error outputs as well (e.g. `package pgdg-redhat96-9.6-3.noarch is already installed` on the `sudo rpm -ivh` step).  Do you feel strongly about including that flag for this script?  If so lets make it work.
(note: if there are obvious or trivial ways to handle the type of example I just provided please swoop in and make the adjustment)

File: rhel-install.sh (hunk 5 above)
Comment (slifty) [inline rhel-install.sh]: Happy to combine where possible -- I do think it is worth separating into dependency blocks; for instance, if we stop caring about mailcatcher, it would be nice to know which yum installs were part of that particular dependency.  Thoughts?

File: rhel-install.sh (hunk 7 above)
Comment (slifty) [inline rhel-install.sh]: No good reason for sleeping -- I was just sleepy myself at the point of writing and was hoping someone would have a suggestion for a better option.  The patch you submitted is much appreciated.

File: rhel-install.sh (hunk 4 above)
Comment (slifty) [inline rhel-install.sh]: I believe the reason I went with direct sql is that createuser requires interaction to define the password; I believe there are ways around that problem using redirection, but it might decrease the legibility of the script.

File: rhel-install.sh (hunk 1 above)
Comment (jasonaowen) [inline rhel-install.sh]: I mean, I feel strongly enough about this to push at least little. :) I could probably be convinced it's not a big deal, by saying it has a limited purpose or a limited scope, but if we're going to distribute this and say "here's how to install" we should make sure it handles errors gracefully, right?

For the `rpm` thing, perhaps [using `--upgrade` instead of `--install` would address that error](https://stackoverflow.com/a/31082948/25637). Are there others?

Comment (slifty) [inline rhel-install.sh]: Not a push at all!  `set -e` is in there, I just wanted to raise the concern that the current script might have cases where it "errors" even when it's not much of an error.  I'll swap in --upgrade; I don't know if there are other cases or not, but since this script has really only been tested on clean OS installs I think it's fair to say that it is only currently supported on clean OS installs at the moment.

**Comment by jvasile** [comment]
+1 to all of Slifty's recent work on this

File: rhel-install.sh (hunk 1 above)
Comment (jvasile) [inline rhel-install.sh]: I'm ok with "only currently supported on clean OS installs at the moment".  If we deliver it with that caveat (and maybe write that somewhere in documentation), can we add "set -e" and merge?  By my count, that's the only outstanding issue preventing merge.
